error: cannot convert 'struct array' to 'int'

### Describe the bug

V throws an error when I try to access a certain function in an array of functions given as an argument.

### Expected Behavior

Should compile properly and let me access the function I want

### Current Behavior
```
==================
/tmp/v_1000/main.3539769618563317589.tmp.c:12291: error: cannot convert 'struct array' to 'int'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.
```

### Reproduction Steps

```v
module main

fn max_speed(functions []fn (int) int, g fn (int) int, end_time int) int {
	mut s := [0]
	mut gears := [0]

	for t in 0..end_time {
		stay_in_gear_speed := if t-1 >= 0 {functions[gears[t-1]](s[t-1])} else {0}
		shift_gear_speed := if t-2 >= 0 {functions[gears[t-2]+1](g(s[t-2]))} else {0}
		if stay_in_gear_speed > shift_gear_speed {
			gears[t] = gears[t-1]
			s << stay_in_gear_speed
		} else {
			gears[t] = gears[t-2] + 1
			s << shift_gear_speed
		}
	}
	return s[end_time]
}

fn first_gear(speed int) int {
	return speed + 1
}

fn second_gear(speed int) int {
	return speed + 1
}

fn third_gear(speed int) int {
	return speed + 1
}

fn decelerate(speed int) int {
	return speed - 3
}

fn main() {
	funcs := [first_gear, second_gear, third_gear]
	println(max_speed(funcs, decelerate, 3))
}
```

### Possible Solution

I tried commenting out the whole for loop and just hard-coding `println(functions[0](s[0]))` -- that's fine. But once I do `println(functions[gears[0]](s[0]))`, it throws the same error.

### Additional Information/Context

Potentially related to https://github.com/vlang/v/issues/15232 ? But in my case I am not passing a slice where a single value is expected.

### V version

V 0.3.3 b1ed1d3

### Environment details (OS name and version, etc.)

OS: linux, Ubuntu 22.04.2 LTS
Processor: 16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P
CC version: cc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0

getwd: /home/neniu/Documents/Gitlab/algos
vmodules: /home/NAME/.vmodules
vroot: /home/NAME/v
vexe: /home/NAME/v/v
vexe mtime: 2023-02-22 03:28:32
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.3 58e150d.b1ed1d3

Git version: git version 2.34.1
Git vroot status: weekly.2023.08-8-gb1ed1d3b
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

--- a/vlib/v/gen/c/index.v
+++ b/vlib/v/gen/c/index.v
@@ -234,6 +234,7 @@ fn (mut g Gen) index_of_array(node ast.IndexExpr, sym ast.TypeSymbol) {
 		}
 	} else {
 		is_direct_array_access := g.is_direct_array_access || node.is_direct
+		is_fn_index_call := g.is_fn_index_call && elem_sym.info is ast.FnType
 		// do not clone inside `opt_ok(opt_ok(&(string[]) {..})` before returns
 		needs_clone := info.elem_type == ast.string_type_idx && g.is_autofree && !(g.inside_return
 			&& g.fn_decl != unsafe { nil } && g.fn_decl.return_type.has_flag(.option))
@@ -257,7 +258,7 @@ fn (mut g Gen) index_of_array(node ast.IndexExpr, sym ast.TypeSymbol) {
 			if needs_clone {
 				g.write('/*2*/string_clone(')
 			}
-			if g.is_fn_index_call {
+			if is_fn_index_call {
 				if elem_sym.info is ast.FnType {
 					g.write('((')
 					g.write_fn_ptr_decl(&elem_sym.info, '')
@@ -297,13 +298,13 @@ fn (mut g Gen) index_of_array(node ast.IndexExpr, sym ast.TypeSymbol) {
 			g.write('data)[')
 			g.expr(node.index)
 			g.write(']')
-			if g.is_fn_index_call {
+			if is_fn_index_call {
 				g.write(')')
 			}
 		} else {
 			g.write(', ')
 			g.expr(node.index)
-			if g.is_fn_index_call {
+			if is_fn_index_call {
 				g.write(')))')
 			} else {
 				g.write('))')

--- a/vlib/v/tests/builtin_arrays/array_of_fns_index_call_with_embeded_array_call_test.v
+++ b/vlib/v/tests/builtin_arrays/array_of_fns_index_call_with_embeded_array_call_test.v
@@ -53,7 +53,7 @@ fn decelerate(speed int) int {
 
 fn test_array_of_fns_index_call_with_embeded_array_call() {
 	funcs := [first_gear, second_gear, third_gear]
-	
+
 	speed1 := max_speed(funcs, decelerate, 3)
 	println(speed1)
 	assert speed1 == 1

--- a/vlib/v/tests/builtin_arrays/array_of_fns_index_call_with_embeded_array_call_test.v
+++ b/vlib/v/tests/builtin_arrays/array_of_fns_index_call_with_embeded_array_call_test.v
@@ -1,0 +1,64 @@
+fn max_speed(functions []fn (int) int, g fn (int) int, end_time int) int {
+	mut s := [0]
+	mut gears := []int{len: end_time, init: 0}
+
+	for t in 0 .. end_time {
+		stay_in_gear_speed := if t - 1 >= 0 { functions[gears[t - 1]](s[t - 1]) } else { 0 }
+		shift_gear_speed := if t - 2 >= 0 { functions[gears[t - 2] + 1](g(s[t - 2])) } else { 0 }
+		if stay_in_gear_speed > shift_gear_speed {
+			gears[t] = if t - 1 >= 0 { gears[t - 1] } else { 0 }
+			s << stay_in_gear_speed
+		} else {
+			gears[t] = if t - 2 >= 0 { gears[t - 2] } else { 0 } + 1
+			s << shift_gear_speed
+		}
+	}
+	return s[end_time]
+}
+
+@[direct_array_access]
+fn max_speed_direct_access(functions []fn (int) int, g fn (int) int, end_time int) int {
+	mut s := [0]
+	mut gears := []int{len: end_time, init: 0}
+
+	for t in 0 .. end_time {
+		stay_in_gear_speed := if t - 1 >= 0 { functions[gears[t - 1]](s[t - 1]) } else { 0 }
+		shift_gear_speed := if t - 2 >= 0 { functions[gears[t - 2] + 1](g(s[t - 2])) } else { 0 }
+		if stay_in_gear_speed > shift_gear_speed {
+			gears[t] = if t - 1 >= 0 { gears[t - 1] } else { 0 }
+			s << stay_in_gear_speed
+		} else {
+			gears[t] = if t - 2 >= 0 { gears[t - 2] } else { 0 } + 1
+			s << shift_gear_speed
+		}
+	}
+	return s[end_time]
+}
+
+fn first_gear(speed int) int {
+	return speed + 1
+}
+
+fn second_gear(speed int) int {
+	return speed + 1
+}
+
+fn third_gear(speed int) int {
+	return speed + 1
+}
+
+fn decelerate(speed int) int {
+	return speed - 3
+}
+
+fn test_array_of_fns_index_call_with_embeded_array_call() {
+	funcs := [first_gear, second_gear, third_gear]
+	
+	speed1 := max_speed(funcs, decelerate, 3)
+	println(speed1)
+	assert speed1 == 1
+
+	speed2 := max_speed_direct_access(funcs, decelerate, 3)
+	println(speed2)
+	assert speed2 == 1
+}

